### DISCUSSION TITLE
new packages being added: ParU and LAGraph

### DISCUSSION THREAD

**Discussion by DrTimothyAldenDavis** [DrTimothyAldenDavis/SuiteSparse#448] [Announcements]
I'm in the process of adding two new packages to SuiteSparse:

- **LAGraph**: a graph library based on GraphBLAS, whose primary repository is https://github.com/GraphBLAS/LAGraph , in collaboration with many co-authors. See https://github.com/DrTimothyAldenDavis/SuiteSparse/tree/dev2/LAGraph .  LAGraph co-authors include @sei-smcmillan, @pcostanza, @eriknw, @michelp, @szarnyasg and many more.  The stable release of LAGraph v1.0.2 has been added, but it's under active development.  Those changes will primarily affect the source, not the cmake build system, however.

- **ParU**: a parallel unsymmetric multifrontal method, in collaboration with @Aznaveh , from https://github.com/Aznaveh/SuiteSparse-official .  It's much faster than UMFPACK.  See https://github.com/DrTimothyAldenDavis/SuiteSparse/tree/dev2/ParU .

ParU has been in a fork of SuiteSparse and its current cmake system had been integrated with SuiteSparse v6.0.2, and needs a cmake build system that fits with the rest of the current version SuiteSparse.  LAGraph needs the same, but it also needs to remain viable as its own stand-alone repo.  That may mean the the SuiteSparse/LAGraph and https://github.com/GraphBLAS/LAGraph will diverge slightly (only in its CMakeLists.txt), but ideally they can be synchronized together so that a single cmake build system can work in both contexts.  In either case, the source code of both will be the same.

@mmuetzel : If you like, I can take a first pass at updating the cmake of these two packages, based on your many awesome updates to the cmake build systems for all the other packages in SuiteSparse.  I should likely wait until your pass at the SuiteSparse cmake is done.  Or if you feel adventurous, please feel free!

**Comment by mmuetzel** [comment]
I'll be happy to give the new packages a pass.
As a first step, I'd like to add these packages to the CI to help avoid breaking things on the way.
Would it be possible to merge #450 early?

**Reply by DrTimothyAldenDavis** [reply]
Thanks!  I've merged in #450, and the others.  There are 2 remaining PRs but I think both aren't ready yet, right?

**Reply by mmuetzel** [reply]
Yeah. I'm not sure if the approach in #426 can work in principle to resolve the task described in #425.
Installing several versions of SuiteSparse as completely independent packages might be better...
But in the end, the maintainer of the packages in the distro will decide what's best for them.

**Comment by mmuetzel** [comment]
Running the demos built for ParU seems to be failing intermittently in the CI with a segmentation fault.
Since it is succeeding most of the time but segfaulting sometimes, it could be some kind of non-initialized variable or a threading issue (or something else).
I haven't looked at the sources of the demos or ParU itself in any detail. Just thought I'd let you know that there might be something in the woods...

**Reply by mmuetzel** [reply]
See also #474.

**Comment by mmuetzel** [comment]
The new projects don't seem to build with OpenMP using a MSVC toolchain (in CI). Afaict, that toolchain supports OpenMP 2.0. Maybe, the new projects try to use features that were added in newer versions of OpenMP?

**Reply by mmuetzel** [reply]
See also #462. Fixed in #463.

**Comment by DrTimothyAldenDavis** [comment]
Looking good so far.  Once the cmake build system is polished, I'll need to take a full pass at ParU to review the code, and to fix the issue where it's relying on umf_internal.h and amd_internal.h.

**Comment by DrTimothyAldenDavis** [comment]
ParU 0.1.0 and LAGraph 1.1.0 now appear in SuiteSparse 7.4.0.beta1, as a pre-release.
